File: MATLAB/iris_py.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%	Author(s): C. Nicolas Barati nicobarati@rice.edu 
%		Rahman Doost-Mohamamdy: doost@rice.edu
%
%---------------------------------------------------------------------
% Copyright (c) 2018-2019, Rice University
% RENEW OPEN SOURCE LICENSE: http://renew-wireless.org/license
% ---------------------------------------------------------------------
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

classdef iris_py < handle
    % This class serves as an interface between matlab and
    % iris python driver. Matlab calls the methods here and iris_class calls
    % iris' python functions.
    
    properties
        sdr_params;
        % pyhton object array. This array decribes 1 Iris board or 
        % a collection of Iris boards that belong to the same entity.E.g., a BS.
        py_obj_array = {};
        use_hub = 0;
        py_obj_hub;
        % Parameters to feed python (pun very intended!)
        serial_ids = [];
        n_sdrs = 0;         % number of Iris boards in a chain
		sample_rate = 0;
		tx_freq = 0;
		rx_freq = 0;
		bw = 0;
		tx_gain = 0;
		rx_gain = 0;
		n_samp = 0;
        tdd_sched = "";
        n_zpad_samp = 0;    %number of zero-padding samples
        n_frame = 10;
        is_bs = 1;
    end
    
    methods
        function obj = iris_py(sdr_params, hub_id) % User must put 0 in hub_params in case hub is not used!
            if nargin > 0
                if ~isempty(hub_id)
                    disp('Using hub with ID: ');
                    disp(hub_id)
                    id_str = convertStringsToChars(hub_id);
                    obj.py_obj_hub = py.hub_py.Hub_py( pyargs('serial_id',id_str));
                    obj.use_hub = 1;
                end
                
                obj.sdr_params = sdr_params;
                
                obj.serial_ids = sdr_params.id; 
                obj.n_sdrs = sdr_params.n_sdrs;
                obj.sample_rate = sdr_params.sample_rate;
                obj.tx_freq = sdr_params.txfreq;
                obj.rx_freq = sdr_params.rxfreq;
                obj.tx_gain = sdr_params.txgain;
                obj.rx_gain = sdr_params.rxgain;
                obj.n_samp = sdr_params.n_samp;
                obj.n_frame = sdr_params.n_frame;
                obj.tdd_sched = sdr_params.tdd_sched; % This is an array
                obj.n_zpad_samp = sdr_params.n_zpad_samp;
                
                for ipy=1:obj.n_sdrs
                    id_str = convertStringsToChars(obj.serial_ids(ipy));
                    py_obj = py.iris_py.Iris_py( pyargs('serial_id',id_str,...
                        'tx_freq', obj.tx_freq, 'rx_freq', obj.rx_freq,...
                        'tx_gain',obj.tx_gain,'rx_gain',obj.rx_gain,...
                        'sample_rate',obj.sample_rate, 'n_samp', obj.n_samp) );
                    
                    obj.py_obj_array(ipy,:) = {py_obj};
                end
            end
        end

        function sdr_set_n_frame(obj, n_frame)
            obj.n_frame = n_frame;
        end
        
        % NB: Need to redefine py_obj as array to be parsed in all the
        % functions that follow
        function sdrtrigger(obj)
            if ~obj.use_hub
                obj.py_obj_array{1}.set_trigger(); % The trigger is set only on the fist node
            else
                obj.py_obj_hub.set_trigger();
                disp('triggering FAROS hub...')
            end
        end
        
         function sdrsync(obj)
            if ~obj.use_hub
                obj.py_obj_array{1}.sync_delays();
            else
                obj.py_obj_hub.sync_delays();
            end
         end
         
         function sdr_setcorr(obj)
             for ipy = 1:obj.n_sdrs
                 obj.py_obj_array{ipy}.set_corr(); 
             end
         end
         
         function sdr_unsetcorr(obj)
             for ipy = 1:obj.n_sdrs
                 obj.py_obj_array{ipy}.unset_corr(); 
             end
         end
         
         function sdr_configgainctrl(obj)
             for ipy = 1:obj.n_sdrs
                 obj.py_obj_array{ipy}.config_gain_ctrl(); 
             end
         end
         
         function sdr_setupbeacon(obj)
             % Assume Beacon only from the first Iris board in the BS array
             for ipy = 1:obj.n_sdrs
                 obj.py_obj_array{ipy}.burn_beacon();
             end
         end
         
         function set_tddconfig(obj, is_bs, tdd_sched)
             obj.is_bs = is_bs;
             sched  = convertStringsToChars(tdd_sched);
             for ipy = 1:obj.n_sdrs
                 obj.py_obj_array{ipy}.config_sdr_tdd( pyargs('tdd_sched', sched, ...
                     'is_bs', is_bs, 'prefix_len', obj.n_zpad_samp, 'max_frames', 1));
             end
         end

         function set_tddconfig_single(obj, is_bs, tdd_sched, index)
             sched  = convertStringsToChars(tdd_sched);
             obj.py_obj_array{index}.config_sdr_tdd( pyargs('tdd_sched', sched, ...
                 'is_bs', is_bs, 'prefix_len', obj.n_zpad_samp, 'max_frames', 1));
         end
         
         function sdrrxsetup(obj)
            for ipy = 1:obj.n_sdrs
                obj.py_obj_array{ipy}.setup_stream_rx();
            end
         end
        
         function sdr_activate_rx(obj)
           for ipy=1:obj.n_sdrs
               obj.py_obj_array{ipy}.activate_stream_rx();
           end
         end
        
        %Assume same data is Tx-ed from all memebers of the array
        function sdrtx(obj, data)
            for ipy = 1:obj.n_sdrs
                obj.py_obj_array{ipy}.burn_data( pyargs('data_r', real(data), 'data_i', imag(data)) );
            end
        end

        function sdrtx_single(obj, data, index)
            obj.py_obj_array{index}.burn_data( pyargs('data_r', real(data), 'data_i', imag(data)) );
        end
        
        % Read n_frame x n_samp data
        function [data, len] = sdrrx(obj, n_samp)
            data_raw = zeros(obj.n_sdrs, obj.n_frame * n_samp);  % Change this to max frame!
            
            for jf=1:obj.n_frame
                %trigger base station
                if obj.is_bs
                    if ~obj.use_hub
                        obj.py_obj_array{1}.set_trigger(); % The trigger is set only on the fist node
                    else
                        obj.py_obj_hub.set_trigger();
                        disp('triggering FAROS hub...')
                    end
                end
                
                for ipy = 1:obj.n_sdrs
                    rcv_data = obj.py_obj_array{ipy}.recv_stream_tdd();
                    data_raw(ipy, (jf-1)*n_samp + 1: jf*n_samp) = double( py.array.array( 'd',py.numpy.nditer( py.numpy.real(rcv_data) ) ) ) + ...
                        1i*double( py.array.array( 'd',py.numpy.nditer( py.numpy.imag(rcv_data) ) ) );
                end
            end
            data = obj.get_best_frame(data_raw.', n_samp);
            len = length(data);
        end
        
        
        function sdr_close(obj)
            for ipy = 1:obj.n_sdrs
                obj.py_obj_array{ipy}.close();
                delete(obj.py_obj_array{ipy});
            end
        end
        
        function [data] = get_best_frame(obj, data_frame, n_samp)
            % FD LTS
            lts_f = [0 1 -1 -1 1 1 -1 1 -1 1 -1 -1 -1 -1 -1 1 1 -1 ...
                -1 1 -1 1 -1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 1 1 -1 -1 ...
                1 1 -1 1 -1 1 1 1 1 1 1 -1 -1 1 1 -1 1 -1 1 1 1 1].';
            % TD LTS
            lts_t = ifft(lts_f, 64);
            % Add CP and a second TD LTS
            lts_cp = lts_t(33:end);
            lts_t = [lts_cp; lts_t; lts_t];
            
            % Correlation through filtering
            lts_corr_dr = zeros(size(data_frame));
            unos = ones(size(lts_t));
            a = 1;
            for nc =1:size(data_frame,2)
                v0 = filter(flipud(conj(lts_t)),a,data_frame(:,nc));
                v1 = filter(unos,a,abs(data_frame(:,nc)).^2);
                lts_corr_dr(:,nc) = (abs(v0).^2)./v1; % normalized correlation
            end
            % Sum accross antennas
            lts_corr_sum = sum(lts_corr_dr,2);
            % zero non-peaks:
            lts_corr_sum(lts_corr_sum < 0.2) = 0.001;
            
            % Assume peak in the first 500 samples
            lts_corr_frm = reshape(lts_corr_sum, [], obj.n_frame);
            
            if obj.n_sdrs == 1 && length(lts_corr_frm >= 300)
                lts_corr_frm = lts_corr_frm(1:300,:);
            elseif (obj.n_sdrs > 1) && length(lts_corr_frm >= 420)
                lts_corr_frm = lts_corr_frm(1:420,:);
            end
            save my_data.mat lts_corr_dr lts_corr_frm
            
            % Avg corr value per frame
            frm_avg_corr = sum(lts_corr_frm,1)./obj.n_sdrs
            % Take index of maximum corr. value
            [max_corr, m_idx] = max(frm_avg_corr);
           
            % Reshape data frame to n_samp-by-n_antenna-by-n_frame
            data_split  = zeros(obj.n_samp, obj.n_sdrs,obj.n_frame);
            for nf = 1:obj.n_frame
                strt_idx = (nf-1)*obj.n_samp +1;
                end_idx = nf*obj.n_samp;
                data_split(:,:,nf) = data_frame(strt_idx :end_idx ,:);
            end
            %data_frame = reshape(data_frame,n_samp, [], obj.n_frame );        
    
            % Return the frame with the highest value 
            data = data_split(:,:,m_idx).';
            fprintf('Returning frame number %d with max mean correlation = %f \n',m_idx,max_corr);
        end
        
    end
end

File: MATLAB/iris_py.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-
""" 
        Author(s): C. Nicolas Barati nicobarati@rice.edu 
                Rahman Doost-Mohamamdy: doost@rice.edu
                Oscar Bejarano: obejarano@rice.edu

---------------------------------------------------------------------
        Copyright (c) 2018-2019, Rice University
        RENEW OPEN SOURCE LICENSE: http://renew-wireless.org/license
---------------------------------------------------------------------
"""

import sys
from argparse import ArgumentParser
import SoapySDR
from SoapySDR import *  # SOAPY_SDR_ constants
from optparse import OptionParser
import numpy as np
import time
import os
import math
import json
import threading
import matplotlib.pyplot as plt
import hub_py
from hub_py import *

# Global variables
corr_threshold = 1

#preambles_bs = generate_training_seq(preamble_type='gold_ifft', seq_length=128, cp=0, upsample=1)
#beacon = preambles_bs[0, :]*.5
#np.savetxt('beacon.txt', np.column_stack([beacon.real, beacon.imag]))

# Read beacon from file
bcn_real, bcn_imag = np.loadtxt('beacon.txt', unpack=True)
beacon = bcn_real + 1j * bcn_imag


#Generate LTS seq:
def gen_lts(seq_length=128, cp=0, upsample=1):
    # Generate 802.11 LTS preamble
    lts_freq = np.array([
        0, 0, 0, 0, 0, 0, 1, 1, -1, -1, 1, 1, -1, 1, -1, 1, 1, 1, 1, 1, 1, -
        1, -1, 1, 1, -1, 1, -1, 1, 1, 1, 1, 0, 1, -1, -1, 1,
        1, -1, 1, -1, 1, -1, -1, -1, -1, -1, 1, 1, -1, -1, 1, -1, 1, -1, 1, 1, 1, 1, 0, 0, 0, 0, 0])
    up_zeros = np.zeros(len(lts_freq) // 2 * (upsample - 1))
    lts_freq_up = np.concatenate((up_zeros, lts_freq, up_zeros))
    signal = np.fft.ifft(np.fft.ifftshift(lts_freq_up))
    # Now affix the cyclic prefix
    # could use tile...
    sequence = np.concatenate((signal[len(signal) - cp:], signal, signal))
    return sequence, lts_freq


def cfloat2uint32(arr, order='IQ'):
    """
    Convert floating point iq values to uint32 (FPGA format)
    ARGS:
    - arr: Data array
    - order: Whether it is IQ or QI
    RETURNS:
    - corresponding uint32 value
    """
    arr_i = (np.real(arr) * 32767).astype(np.uint16)
    arr_q = (np.imag(arr) * 32767).astype(np.uint16)
    if order == 'IQ':
        return np.bitwise_or(arr_q, np.left_shift(arr_i.astype(np.uint32), 16))
    else:
        return np.bitwise_or(arr_i, np.left_shift(arr_q.astype(np.uint32), 16))


#######################################
####### SDR Class:              #######
#######################################

class Iris_py:
        '''
                Iris python class. To act as a middle layer between Matlab classes and SoapySDRs
        '''

        def __init__(self,
              serial_id=None,
              tx_freq=None,
              rx_freq=None,
              tx_gain=None,
              rx_gain=None,
              bw=None,
              sample_rate=None,
              n_samp=None,                              # Total number of samples, including zero-pads
              both_channels=False,
              agc_en=False,
              ):

                if serial_id is not None:
                        self.sdr = SoapySDR.Device(
                            dict(driver="iris", timeout="1000000", serial=serial_id))
                        self.serial_id = serial_id
                else:
                        self.sdr = None

                self.sample_rate = sample_rate

                if n_samp is not None:
                    self.n_samp = int(n_samp)

                self.agc_en = agc_en
                self.both_channels = both_channels
                self.max_frames = 1

                ### Setup channel rates, ports, gains, and filters ###
                info = self.sdr.getHardwareInfo()
                for chan in [0, 1]:

                        #Tx:
                        if sample_rate is not None:
                                self.sdr.setSampleRate(SOAPY_SDR_TX, chan, sample_rate)
                        if bw is not None:
                                self.sdr.setBandwidth(SOAPY_SDR_TX, chan, bw)
                        else:
                                self.sdr.setBandwidth(SOAPY_SDR_TX, chan, 2.5*sample_rate)
                        if tx_gain is not None:
                                self.sdr.setGain(SOAPY_SDR_TX, chan, min(tx_gain, 81.0))
                        if tx_freq is not None:
                                self.sdr.setFrequency(SOAPY_SDR_TX, chan, 'RF', tx_freq - .75*sample_rate)
                                self.sdr.setFrequency(SOAPY_SDR_TX, chan, 'BB', .75*sample_rate)

                        #print("Set TX frequency to %f" % self.sdr.getFrequency(SOAPY_SDR_TX, chan))
                        #self.sdr.setAntenna(SOAPY_SDR_TX, chan, "TRX")
                        #self.sdr.setGain(SOAPY_SDR_TX, chan, 'ATTN', -6)

                        #Rx:
                        if sample_rate is not None:
                                self.sdr.setSampleRate(SOAPY_SDR_RX, chan, sample_rate)
                        if bw is not None:
                                self.sdr.setBandwidth(SOAPY_SDR_RX, chan, bw)
                        else:
                                self.sdr.setBandwidth(SOAPY_SDR_RX, chan, 2.5*sample_rate)
                        if rx_gain is not None:
                                self.sdr.setGain(SOAPY_SDR_RX, chan, rx_gain)
                        if rx_freq is not None:
                                self.sdr.setFrequency(SOAPY_SDR_RX, chan, 'RF', rx_freq - .75*sample_rate)
                                self.sdr.setFrequency(SOAPY_SDR_RX, chan, 'BB', .75*sample_rate)

                        self.sdr.setAntenna(SOAPY_SDR_RX, chan, "TRX")

                        if self.agc_en:
                                self.sdr.setGain(SOAPY_SDR_RX, chan, 100)  # high gain value
                        else:
                                self.sdr.setGain(SOAPY_SDR_RX, chan, rx_gain)

                        self.sdr.setDCOffsetMode(SOAPY_SDR_RX, chan, True)

                self.tx_stream = None  # Burst mode

                self.sdr.writeSetting("RESET_DATA_LOGIC", "")

                if not self.both_channels:
                        self.sdr.writeSetting(SOAPY_SDR_RX, 1, 'ENABLE_CHANNEL', 'false')
                        self.sdr.writeSetting(SOAPY_SDR_TX, 1, 'ENABLE_CHANNEL', 'false')

        # Set trigger:
        def set_trigger(self):
                self.sdr.writeSetting("TRIGGER_GEN", "")

        def set_corr(self):
                '''enable the correlator, with inputs from adc'''
                self.sdr.writeSetting("CORR_START", "A")

        def unset_corr(self):
                corr_conf = {"corr_enabled": False}
                self.sdr.writeRegister("IRIS30", 60, 
                    self.sdr.readRegister("IRIS30", 60) & 0xFFFE)

        def config_sdr_tdd(self, is_bs=True, tdd_sched="G", prefix_len=0, max_frames=1):
                '''Configure the TDD schedule and functionality when unchained. Set up the correlator.'''
                global corr_threshold, beacon
                coe = cfloat2uint32(np.conj(beacon), order='QI')
                self.tdd_sched = tdd_sched
                self.max_frames = int(max_frames)
                if bool(is_bs):
                        conf_str = {"tdd_enabled": True,
                            "frame_mode": "free_running",
                            "symbol_size": self.n_samp,
                            "max_frame": max_frames,
                            "beacon_start": prefix_len,
                            "beacon_stop": prefix_len + len(beacon),
                            "frames": [self.tdd_sched]}
                        self.sdr.writeSetting("TDD_CONFIG", json.dumps(conf_str))
                        print("TDD schedule of BS node {}: {}".format(self.serial_id, tdd_sched))
                else:
                        conf_str = {"tdd_enabled": True,
                            "frame_mode": "triggered",
                            "symbol_size": self.n_samp,
                            "frames": [self.tdd_sched]}
                        self.sdr.writeSetting("TDD_CONFIG", json.dumps(conf_str))

                        #Correlator stuff:
                        corr_conf = {"corr_enabled": True,
                                     "corr_threshold": corr_threshold}
                        self.sdr.writeSetting(
                            "CORR_CONFIG", json.dumps(corr_conf))
                        if coe is not None:
                                self.sdr.writeRegisters(
                                    "CORR_COE", 0, coe.tolist())
                        else:
                                print("No coe was passed into config_sdr_tdd() \n")

                        # DEV: ueTrigTime = 153 (prefix_length=0), CBRS: ueTrigTime = 235 (prefix_length=82), tx_advance=prefix_length, corr delay is 17 cycles
                        ueTrigTime = len(beacon) + 200
                        sf_start = int(ueTrigTime//(self.n_samp))
                        sp_start = int(ueTrigTime % (self.n_samp))
                        print("config_sdr_tdd: UE starting symbol and sample count (%d, %d)" %
                              (sf_start, sp_start))
                        # make sure to set this after TDD mode is enabled "writeSetting("TDD_CONFIG", ..."
                        self.sdr.setHardwareTime(SoapySDR.ticksToTimeNs(
                            (sf_start << 16) | sp_start, self.sample_rate), "TRIGGER")
                        print("TDD schedule of UE node {}: {}".format(self.serial_id, tdd_sched))

                self.sdr.writeSetting("TX_SW_DELAY", str(30))
                self.sdr.writeSetting("TDD_MODE", "true")

        def config_gain_ctrl(self):
                tpc_conf = {"tpc_enabled": False}
                self.sdr.writeSetting("TPC_CONFIG", json.dumps(tpc_conf))
                agc_conf = {"agc_enabled": self.agc_en}
                self.sdr.writeSetting("AGC_CONFIG", json.dumps(agc_conf))

        def sync_delays(self):
                # NB: This should be called for the BS
                self.sdr.writeSetting("SYNC_DELAYS", "")

        # Setup and activate RX streams
        def setup_stream_rx(self):
                print("Setting up RX stream \n")
                self.rx_stream = self.sdr.setupStream(SOAPY_SDR_RX, SOAPY_SDR_CF32, [0, 1])

        def activate_stream_rx(self, flags=0):
                r1 = self.sdr.activateStream(self.rx_stream, flags, 0)
                if r1 < 0:
                    print("Problem activating stream\n")

        def burn_beacon(self):
                '''Write beacon to the FPGA ram'''
                buf_a = cfloat2uint32(beacon, order='QI')
                beacon_weights = [[1, 1], [1, 1]]
                self.sdr.writeRegisters("BEACON_RAM", 0, buf_a.tolist())
                self.sdr.writeRegisters(
                    "BEACON_RAM_WGT_A", 0, beacon_weights[0])
                self.sdr.writeRegisters(
                    "BEACON_RAM_WGT_B", 0, beacon_weights[1])
                self.sdr.writeSetting("BEACON_START", str(2))

        # Write data to FPGA RAM
        def burn_data(self, data_r, data_i=None, replay_addr=0):
                '''Write data to FPGA RAM. A pilot for example. Need to compose a complex vector out of data_r and data_i'''

                if data_i is not None:
                    # hack for matlab...
                    data = np.asarray(data_r, dtype=np.complex64) + 1.j * \
                        np.asarray(data_i, dtype=np.complex64)

                buf_a = cfloat2uint32(data, order='QI')
                print("burnt data on to node {} FPGA RAM".format(self.serial_id))

                self.sdr.writeRegisters("TX_RAM_A", replay_addr, buf_a.tolist())
                #self.sdr.writeRegisters("TX_RAM_B", replay_addr, buf_b.tolist() )

        def recv_stream_tdd(self):
                '''Read an incoming stream.'''
                max_frames = int(self.max_frames)
                in_len = int(self.n_samp)
                wave_rx_a = np.zeros((in_len), dtype=np.complex64)
                wave_rx_b = np.zeros((in_len), dtype=np.complex64)
                rx_frames_a = np.zeros((in_len*max_frames), dtype=np.complex64)

                n_R = self.tdd_sched.count("R")  # How many Read frames in the tdd schedule
                print("n_samp is: %d  \n" % self.n_samp)

                for m in range(max_frames):
                    for k in range(n_R):
                        r1 = self.sdr.readStream(
                            self.rx_stream, [wave_rx_a, wave_rx_b], int(self.n_samp))
                        print("reading stream: ({})".format(r1))
                    rx_frames_a[m*in_len: (m*in_len + in_len)] = wave_rx_a

                return(rx_frames_a)

        def close(self):
                '''Cleanup streams. Rest SDRs'''
                print("Cleaning up streams")
                if self.sdr is not None:
                        if self.tx_stream is not None:
                                self.sdr.deactivateStream(self.tx_stream)
                                self.sdr.closeStream(self.tx_stream)
                                self.tx_stream = None

                        if self.rx_stream is not None:
                                self.sdr.deactivateStream(self.rx_stream)
                                self.sdr.closeStream(self.rx_stream)
                                self.rx_stream = None

                        self.sdr.writeSetting("RESET_DATA_LOGIC", "")
                        tdd_conf = {"tdd_enabled": False}
                        self.sdr.writeSetting(
                            "TDD_CONFIG", json.dumps(tdd_conf))
                        agc_conf = {"agc_enabled": False}
                        self.sdr.writeSetting(
                            "AGC_CONFIG", json.dumps(agc_conf))
                        self.sdr = None  # still doesn't release handle... you have to kill python.

                print("Done!")



#Used only in case of testing without Matlab:
if __name__ == '__main__':

        parser = ArgumentParser()
        parser.add_argument("--serial-id1", type=str, dest="serial_id1",
                            help="BS SDR Serial Number, e.g., RF3E000XXX", default=None)
        parser.add_argument("--serial-id2", type=str, dest="serial_id2",
                            help="UE SDR Serial Number, e.g., RF3E000XXX", default=None)
        parser.add_argument("--hub-serial", type=str, dest="hub_serial",
                            help="Hub Serial Number, e.g., FH4XXXXXXX", default=None)
        parser.add_argument("--rate", type=float, dest="rate",
                            help="Sample rate", default=5e6)
        parser.add_argument("--txGain", type=float, dest="txGain",
                            help="Optional Tx gain (dB)", default=40)
        parser.add_argument("--rxGain", type=float, dest="rxGain",
                            help="Optional Rx gain (dB)", default=20)
        parser.add_argument("--freq", type=float, dest="freq",
                            help="Optional Tx freq (Hz)", default=3.6e9)
        parser.add_argument("--bw", type=float, dest="bw",
                            help="Optional filter bw (Hz)", default=None)
        args = parser.parse_args()

        siso_bs = Iris_py(
                serial_id=args.serial_id1,
                sample_rate=args.rate,
                tx_freq=args.freq,
                rx_freq=args.freq,
                bw=args.bw,
                tx_gain=args.txGain,
                rx_gain=args.rxGain,
                n_samp=1024,
        )

        siso_ue = Iris_py(
                serial_id=args.serial_id2,
                sample_rate=args.rate,
                tx_freq=args.freq,
                rx_freq=args.freq,
                bw=args.bw,
                tx_gain=args.txGain,
                rx_gain=args.rxGain,
                n_samp=1024,
        )

        hub = None
        if args.hub_serial is not None:
            hub = Hub_py(args.hub_serial)

        print(siso_ue.sample_rate)
        #Generate signal to send
        nsamps = siso_bs.n_samp
        upsample = 1
        ampl = 1
        nsamps_pad = 82
        n_sym_samp = nsamps + 2*nsamps_pad - 14

        ltsSym, lts_f = gen_lts(cp=32, upsample=1)

        # to comprensate for front-end group delay
        pad1 = np.zeros((nsamps_pad), np.complex64)
        # to comprensate for rf path delay
        pad2 = np.zeros((nsamps_pad-14), np.complex64)

        wb_pilot = np.tile(ltsSym, nsamps//len(ltsSym)).astype(np.complex64)*.5
        wbz = np.zeros((n_sym_samp), dtype=np.complex64)
        wb_pilot1 = np.concatenate([pad1, wb_pilot, pad2])
        wb_pilot2 = wbz  # wb_pilot1 if both_channels else wbz

        print("len_beacon: ")
        print(len(beacon))

        wb_pilot1_r = np.real(wb_pilot1)
        wb_pilot1_i = np.imag(wb_pilot1)

        siso_ue.config_gain_ctrl()

        siso_ue.setup_stream_rx()
        siso_bs.setup_stream_rx()

        siso_bs.config_sdr_tdd(tdd_sched="BGGGGGRG", prefix_len=nsamps_pad)
        siso_ue.config_sdr_tdd(tdd_sched="GGGGGGPG", is_bs=False)

        siso_bs.burn_beacon()
        siso_ue.burn_data(wb_pilot1_r, wb_pilot1_i)

        siso_bs.activate_stream_rx()

        siso_ue.set_corr()
        if hub is not None:
            hub.set_trigger()
        else:
            siso_bs.set_trigger()

        wave_rx_a_bs_mn = siso_bs.recv_stream_tdd()

        freq = args.freq
        print("printing number of frames")
        print("BS 0x%X" % SoapySDR.timeNsToTicks(
            siso_bs.sdr.getHardwareTime(""), freq))
        print("UE 0x%X" % SoapySDR.timeNsToTicks(
            siso_ue.sdr.getHardwareTime(""), freq))

        siso_ue.close()
        siso_bs.close()

        #Test:
        plt.close('all')
        fig = plt.figure(figsize=(20, 8), dpi=100)
        ax1 = fig.add_subplot(2, 1, 1)
        ax2 = fig.add_subplot(2, 1, 2)
        ax1.plot(np.real(wb_pilot), label='pilot i')
        ax1.plot(np.imag(wb_pilot), label='pilot q')
        ax2.plot(np.real(wave_rx_a_bs_mn), label='rx data i')
        ax2.plot(np.imag(wave_rx_a_bs_mn), label='rx data q')
        plt.show()
